swap/src/bitcoin/timelocks.rs

```diff
@@ -55,3 +55,44 @@ pub enum ExpiredTimelocks {
     Cancel,
     Punish,
 }
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(transparent)]
+pub struct CancelTimelock(Timelock);
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(transparent)]
+pub struct PunishTimelock(Timelock);
+
+macro_rules! impl_timelock {
+    ($type:ident) => {
+        impl Add<$type> for BlockHeight {
+            type Output = BlockHeight;
+
+            fn add(self, rhs: $type) -> Self::Output {
+                BlockHeight::new(rhs.0 .0 + self.0)
+            }
+        }
+
+        impl $type {
+            pub const fn new(number_of_blocks: u32) -> Self {
+                Self(Timelock::new(number_of_blocks))
+            }
+        }
+
+        impl From<$type> for u32 {
+            fn from(timelock: $type) -> Self {
+                timelock.0 .0
+            }
+        }
```

Comment:
Such conversions on newtypes take some of their power away.

One way of dealing with this is to try and define the newtype in the same module as where the internals are accessed. In this case, I think only `TxCancel` needs access to the internals. If you would define `CancelTimelock` next to `TxCancel`, you could naturally access the private field without having to expose it publicly.

That would slightly increase the guarantees you get out of the newtype because you can isolate the code that gets "superpowers" and can access internals.

Reply:
A newtype should always enforce its invariants across its full public API. If we can get the internals out and re-construct the newtype from its internals, there is no way we can enforce our invariants.